scheduler cache: Node level aggregated pods' resource summary

hongchaodeng opened the pull request:
This PR provides NodeInfoCache which aggregates pods' resource request and provides node-level resource summary. It's intended to supplant system modeler for efficient lookup in `PodFitsResource`. As we indicated in #18831 that it is the most time-consuming code path and optimizing it would result in 10x improvement. 

The interface defined in the cache is similar to system modeler and its pod lister. Additionally, we also need Update() and GetNodeInfo(). 

We have drawn the state machine explicitly to show the workflow and how it interacts with pod events. 

It would be pretty straight-forward to use this cache into related hooks when state changes happen. That say, when pod is assumed, added, updated, removed. Then we can use the cache to optimize scheduler predicates `PodFitsResources`.

hongchaodeng committed:
scheduler cache: NodeInfoCache for node level aggregated pods' info

hongchaodeng commented:
/cc @xiang90  @davidopp @wojtek-t @timothysc @HaiyangDING @mqliang 

BTW, I won't be available until the day after Christmas.

xiang90 commented on plugin/pkg/scheduler/schedulercache/interface.go:
```diff
@@ -0,0 +1,81 @@
+	RemovePodByKey(key string) error
+	// GetNodeInfo returns aggregated node information for given node name.
+	// If no pod's been aggregated on the node, it returns nil.
+	GetNodeInfo(nodeName string) *NodeInfo
```
Can we move GetNodeInfo to the top of the interface declaration?

xiang90 commented on plugin/pkg/scheduler/schedulercache/node_info.go:
```diff
@@ -0,0 +1,99 @@
+type NodeInfo struct {
+	PodNum            int
+	UsedPorts         map[int]bool
+	RequestedResource *Resource
```
Why do we need a pointer type here?

hongchaodeng replied:
We need to modify it.

xiang90 commented on plugin/pkg/scheduler/schedulercache/node_info.go:
```diff
@@ -0,0 +1,99 @@
+}
+
+// AddPodInfo adds pod information to this NodeInfo.
+func (n *NodeInfo) AddPodInfo(pod *api.Pod) {
```
Just AddPod?

xiang90 commented on plugin/pkg/scheduler/schedulercache/interface.go:
```diff
@@ -0,0 +1,81 @@
+// - Both "Initial" and "Deleted" pods do not actually exist in cache.
+//   In order to differentiate them, we need external request to guarantee
+//   no same pod will be created twice.
+type NodeInfoCache interface {
```
I feel NodeResourceCache is more accurate. We do not actually cache node information, but we cache the resources this node is using.

hongchaodeng replied:
This stores node information in my mind -- how many pods are here, what ports are being used, cpu and mem usage.

Naming is always ambiguous. Let's wait for more comments on this.

xiang90 commented on plugin/pkg/scheduler/schedulercache/node_info.go:
```diff
@@ -0,0 +1,99 @@
+}
+
+// SubPodInfo subtracts pod information to this NodeInfo.
+func (n *NodeInfo) SubPodInfo(pod *api.Pod) {
```
RemovePod?

xiang90 commented on plugin/pkg/scheduler/schedulercache/nodeinfocache/node_info_cache.go:
```diff
@@ -0,0 +1,248 @@
+// This is copied from client/cache.MetaNamespaceKeyFunc .
+// Since pod is ensured to have object, we don't need to take care of the error returned.
+// A more appropriate way is to support Key() method in pod object.
+func podKeyFunc(pod *api.Pod) string {
```
I am worry about this copy actually. How are we going to keep them in sync?

hongchaodeng replied:
I figured that a local copy isn't a good idea.
But we don't want to handle the error returned by `MetaNamespaceKeyFunc`.
How about we have a `mustGetPodKey` that panics if `MetaNamespaceKeyFunc( pod )` returns error.
Another solution I have is that we add `Key()` to pod struct. I'm a little against this because it touches API and comes beyond the scope here.

xiang90 commented on plugin/pkg/scheduler/schedulercache/nodeinfocache/node_info_cache.go:
```diff
@@ -0,0 +1,248 @@
+type nodeInfoCache struct {
+	stop chan struct{}
+
+	mu          sync.Mutex
```
add a comment about this mu? What does it guard?

xiang90 commented on plugin/pkg/scheduler/schedulercache/nodeinfocache/node_info_cache.go:
```diff
@@ -0,0 +1,248 @@
+
+type assumedPod struct {
+	pod            *api.Pod
+	expectToExpire time.Time
```
expectToExpire -> deadline?

xiang90 commented on plugin/pkg/scheduler/schedulercache/nodeinfocache/node_info_cache.go:
```diff
@@ -0,0 +1,248 @@
+	go util.Until(cache.cleanupExpiredAssumedPods, cache.period, cache.stop)
+}
+
+// return nil if no such node is found for given node name.
```
// GetNodeInfo returns cached NodeInfo. It returns nil if no such node if found for given node name.

xiang90 commented on plugin/pkg/scheduler/schedulercache/nodeinfocache/node_info_cache.go:
```diff
@@ -0,0 +1,248 @@
+}
+
+// separate out form time deterministic testsability
+func (cache *nodeInfoCache) assumePod(pod *api.Pod, now time.Time) error {
```
Let's follow the go style of comment. assumePod exists for making test deterministic by taking time as input argument.

xiang90 commented on plugin/pkg/scheduler/schedulercache/nodeinfocache/node_info_cache.go:
```diff
@@ -0,0 +1,248 @@
+
+	key := podKeyFunc(pod)
+	if _, ok := cache.podStates[key]; ok {
+		return fmt.Errorf("Pod state wasn't initial but get assumed. Pod key: %v", key)
```
pod state was't initial but assumed. Pod key... (error message should not be capitalized)

see https://github.com/golang/go/wiki/CodeReviewComments#error-strings

xiang90 commented on plugin/pkg/scheduler/schedulercache/nodeinfocache/node_info_cache.go:
```diff
@@ -0,0 +1,248 @@
+type podState int
+
+const (
+	podBinded podState = iota + 1
```
I feel podAsummed, podAdded, podExpired might be more clear.

xiang90 commented on plugin/pkg/scheduler/schedulercache/nodeinfocache/node_info_cache.go:
```diff
@@ -0,0 +1,248 @@
+		// Pod was expired and then deleted. We should add it back.
+		cache.addPod(pod, key)
+	default:
+		return fmt.Errorf("Pod state wasn't binded or expired but get added. Pod key: %v", key)
```
pod state

xiang90 commented on plugin/pkg/scheduler/schedulercache/nodeinfocache/node_info_cache.go:
```diff
@@ -0,0 +1,248 @@
+	case ok && state == podAdded:
+		cache.updatePod(pod, key)
+	default:
+		return fmt.Errorf("Pod state wasn't added but get updated. Pod key: %v", key)
```
pod state

xiang90 commented on plugin/pkg/scheduler/schedulercache/nodeinfocache/node_info_cache.go:
```diff
@@ -0,0 +1,248 @@
+	case ok && state == podAdded:
+		cache.deletePod(cache.pods[key], key)
+	default:
+		return fmt.Errorf("Pod state wasn't binded, expired, or added but get removed. Pod key: %v", key)
```
pod state

xiang90 commented:
@hongchaodeng How are away from actually putting this cache in use?

hongchaodeng commented:
> How are away from actually putting this cache in use?
I didn't catch that

xiang90 commented:
> Do you mean how this cache is gonna be used?

I mean how much effort would it take to use this cache.

hongchaodeng commented:
> I mean how much effort would it take to use this cache.

It would be pretty straight-forward to use this cache into related hooks when state changes happen. That say, when pod is assumed, added, updated, removed.

Then we can use the cache to optimize scheduler predicates where `PodFitsResources` and `PodFitsHost` are the top two inefficient code paths.

hongchaodeng commented:
@xiang90 
One change I came up lately. I change GetNodeInfo to serve result in callback. In this way, we can better serve results in concurrent environment.

davidopp commented:
@lavalamp

lavalamp commented:
If this is a system state modeler, can the name include something about modeling the system? I'm totally fine with replacing the existing modeler, it was only ever a placeholder. But please actually replace it, don't make a parallel structure.

lavalamp commented on plugin/pkg/scheduler/schedulercache/interface.go:
```diff
@@ -0,0 +1,83 @@
+//                                                |       |
+//                                                |       | Update
+//           Assume                Add            +       |
+// Initial +--------> Binded  +----------------> Added <--+
```
"Bound" is the traditional past-tense of bind.

hongchaodeng replied:
It's changed to "Assumed"

hongchaodeng commented:
Hi @lavalamp , I do agree with you that it would be better to fully combine this cache and system modeler. I want to also point out that the goal here is to make the changes we mentioned in #18831 -- to optimize `PodFitsResource` and `PodFitsHostPort` only. As part of the scalability timeline, we would prefer a fast turnaround for it. Is that OK? I would like to hear your suggestion.

davidopp commented:
> This is the first step of our proposal in #18831 .

What other work do you need for #18831? Just speeding up `PodFitsHostPorts`?

hongchaodeng commented:
Hi @davidopp ,

> What other work do you need for #18831? 

I am expecting to finish the following after this:
- integrate the cache with scheduler for watch-based updates
- pass the cache as an argument to `FitPredicate` and speed up `PodFitsResource` and `PodFitsHostPort`

I have the changes done already. I might choose to split them in one or two PRs based on the size of change. Does that answer your question? Let me know if you prefer it in a single PR. Thanks!

davidopp commented:
> pass the cache as an argument to FitPredicate and speed up PodFitsResource and PodFitsHostPort

Did you mean "pass the cache as an argument to FitPredicate **to** speed up PodFitsResource and PodFitsHostPort" or (what you wrote) "pass the cache as an argument to FitPredicate **and** speed up PodFitsResource and PodFitsHostPort" ?

If the latter, what do you plan to do to speed up PodFitsResource and PodFitsHostPort other than this PR, integrating cache with scheduler for watch-based updates, and passing cache as an argument to FitPredicate?

hongchaodeng commented:
Hi @davidopp ,

> Did you mean "pass the cache as an argument to FitPredicate to speed up PodFitsResource and PodFitsHostPort" or (what you wrote) "pass the cache as an argument to FitPredicate and speed up PodFitsResource and PodFitsHostPort" ?

The former :)
Since we have simple turnaround for `PodFitsHostPort`, it's only `PodFitsResource` that needs speedup.

The change is gonna replace the logic around [CheckPodsExceedingFreeResources](https://github.com/kubernetes/kubernetes/blob/f9a6ac077e653ee9c44b561fbdd9fc14d18308b5/plugin/pkg/scheduler/algorithm/predicates/predicates.go#L321) and use O(1) cache to get the result.

davidopp commented:
OK, sounds good.

hongchaodeng commented:
Updated and removed `UsedPorts` information.

HardySimpson commented on plugin/pkg/scheduler/schedulercache/interface.go:
```diff
@@ -0,0 +1,83 @@
+//                        |               |            |
+//                        |               |            v
+//                        +---------------+------> Deleted
+//                              Remove
```
What's the difference between Assumed and Expired? It seems that Add and Remove works same from them

timothysc commented on plugin/pkg/scheduler/schedulercache/interface.go:
```diff
@@ -0,0 +1,81 @@
+//   no same pod will be created twice.
+type NodeInfoCache interface {
+	// GetNodeInfo returns aggregated node information for given node name.
+	// If no pod's been added on the node, it returns nil.
```
s/pod's/pod has

timothysc commented on plugin/pkg/scheduler/schedulercache/interface.go:
```diff
@@ -0,0 +1,81 @@
+	// AssumePod assumes a pod to be scheduled. The pod's information is aggregated into assigned node.
+	// The implementation might decide the policy to expire/remove the assumed pod before it is confirmed to be scheduled.
+	// After expiration, its information would be subtracted.
+	AssumePod(pod *api.Pod) error
```
I'm terrible at naming, so it's difficult for me to find a good one... but every time I read this the name 'AssumePod' just strikes me as non-intuitive.

hongchaodeng replied:
Assumed is a better state term to differentiate "assumed to be scheduled" and "confirmed it's scheduled".

If it's confusing, how about `AssumePodScheduled`?

timothysc replied:
+1 to ```AssumePodScheduled```

timothysc commented on plugin/pkg/scheduler/schedulercache/nodeinfocache/node_info_cache.go:
```diff
@@ -0,0 +1,249 @@
+import (
+	"sync"
+	"time"
+
```
collapse "fmt" into block above.

timothysc commented on plugin/pkg/scheduler/schedulercache/nodeinfocache/node_info_cache.go:
```diff
@@ -0,0 +1,249 @@
+	if !ok {
+		return nil
+	}
+	if n.PodNum == 0 {
```
why not just collapse conditionals w/ ||

timothysc commented on plugin/pkg/scheduler/schedulercache/node_info.go:
```diff
@@ -0,0 +1,83 @@
+type PodInfo struct {
+	RequestedResource Resource
+}
+
```
general comment on consistency of commenting functions and structs.  Should apply comments to all.

timothysc commented on plugin/pkg/scheduler/schedulercache/nodeinfocache/node_info_cache.go:
```diff
@@ -0,0 +1,245 @@
+	cache.assumedPods[key] = aPod
+	return nil
+}
+
```
see other comment re: comments.

hongchaodeng replied:
These are internal implementation. We only need to write docs for public API and notes on implicit code. Right?

timothysc commented on plugin/pkg/scheduler/schedulercache/nodeinfocache/node_info_cache.go:
```diff
@@ -0,0 +1,245 @@
+
+	// This mutex guards all fields within this cache struct
+	mu          sync.Mutex
+	ttl         time.Duration
```
what exactly is the concrete use case from the scheduler in setting a ttl in the cache?

hongchaodeng replied:
This is related to assumed pods' logic. See https://github.com/kubernetes/kubernetes/blob/15b30b8b099377b1ddd6868aa13a3141b9e2ac28/plugin/pkg/scheduler/modeler.go#L108

In my understanding, this prevents failed or slow detection of unscheduled pods.
See the state machine drawn in this PR, assumed pods need to be confirmed, otherwise it would expire.

timothysc commented on plugin/pkg/scheduler/schedulercache/nodeinfocache/node_info_cache_test.go:
```diff
@@ -0,0 +1,410 @@
+func TestExpirePod(t *testing.T) {
+	nodeName := "node"
+	now := time.Now()
+	ttl := 10 * time.Second
```
Why would a cache not directly get updated from the reflector and use a ttl as shown here?

hongchaodeng replied:
The idea occurred to me.

Let's walk through the current architecture. There are two stores: `ScheduledPodLister` and `assumedPods`. If we can combine these two stores, then it's exactly what I am proposing here -- the scheduler cache.

@lavalamp also mentioned above that we could replace the system modeler instead of having parallel structs.

However, such reconstruction needs a lot more work. This PR is the minimal change that I need to make #18831 work.

I'm pretty open to the scope of the problem and do more work. I'm just not sure what everyone's goal is. I was going to talk to @lavalamp this week since he just came back to bay area.

timothysc replied:
IMHO The codebase has enough duplicated oddities that it's fair to say, "what is there got us working", but may not be what we want going fowards... So rationalizing the structure makes a lot of sense to me.  

+1 to decruft. 

/cc @davidopp

hongchaodeng committed:
AssumePod -> AssumePodScheduled

hongchaodeng commented:
@timothysc 
Changed to `AssumedPodScheduled`

hongchaodeng commented:
After discussing with @lavalamp , I totally agree with him that we should combine system modeler and scheduledPods store into one cache. I will work on it and send out another PR in a couple of days.

timothysc commented:
awesome thx guys!

hongchaodeng commented:
Closing this in favor of #20257.
Will resume the work after that.
